test/units/testsuite-56.sh

```diff
@@ -29,11 +29,13 @@ systemd-run --wait --unit=one -p ExitType=cgroup /tmp/test56-exit-cgroup.sh
 systemd-run --wait --unit=two -p ExitType=cgroup -p ExecCondition=true /tmp/test56-exit-cgroup.sh
 
 # false exec condition: systemd-run should exit immediately with status code: 1
-! systemd-run --wait --unit=three -p ExitType=cgroup -p ExecCondition=false /tmp/test56-exit-cgroup.sh
+systemd-run --wait --unit=three -p ExitType=cgroup -p ExecCondition=false /tmp/test56-exit-cgroup.sh \
+    && { echo 'unexpected success'; exit 1; }
 
 # service should exit uncleanly
 (sleep 1; systemctl kill --signal 9 four) &
-! systemd-run --wait --unit=four -p ExitType=cgroup /tmp/test56-exit-cgroup.sh
+systemd-run --wait --unit=four -p ExitType=cgroup /tmp/test56-exit-cgroup.sh \
+    && { echo 'unexpected success'; exit 1; }
```

Comment:
```
[    7.277975] testsuite-56.sh[466]: + systemd-run --wait --unit=four -p ExitType=cgroup /tmp/test56-exit-cgroup.sh
[    7.278379] testsuite-56.sh[502]: + sleep 1
...
[    7.306050] systemd[1]: Started /tmp/test56-exit-cgroup.sh.
[    7.306423] systemd[505]: four.service: Executing: /tmp/test56-exit-cgroup.sh
[    7.312150] testsuite-56.sh[503]: Running as unit: four.service
[    7.315455] systemd[1]: systemd-journald.service: Got notification message from PID 413 (FDSTORE=1)
[    7.315602] systemd[1]: systemd-journald.service: Added fd 40 (n/a) to fd store.
[    7.316978] test56-exit-cgroup.sh[505]: + disown
[    7.317466] test56-exit-cgroup.sh[506]: + sleep infinity
[    7.318281] test56-exit-cgroup.sh[505]: + sleep infinity
[    7.318652] test56-exit-cgroup.sh[509]: + sleep infinity
...
[    8.281835] testsuite-56.sh[502]: + systemctl kill --signal 9 four
...
[    8.301788] systemd[1]: four.service: Sent signal SIGKILL to main process 505 (bash) on client request.
[    8.302166] systemd[1]: four.service: Sending signal SIGKILL to process 506 (sleep) on client request.
[    8.302559] systemd[1]: four.service: Sending signal SIGKILL to process 507 (bash) on client request.
[    8.302779] systemd[1]: four.service: Sending signal SIGKILL to process 508 (sleep) on client request.
[    8.302988] systemd[1]: four.service: Sending signal SIGKILL to process 509 (sleep) on client request.
[    8.303502] systemd[1]: Received SIGCHLD from PID 506 (sleep).
[    8.303704] systemd[1]: Child 505 (bash) died (code=killed, status=9/KILL)
[    8.303902] systemd[1]: four.service: Child 505 belongs to four.service.
[    8.304151] systemd[1]: Child 506 (sleep) died (code=killed, status=9/KILL)
[    8.304354] systemd[1]: Child 507 (bash) died (code=killed, status=9/KILL)
[    8.304570] systemd[1]: Child 508 (sleep) died (code=killed, status=9/KILL)
[    8.304783] systemd[1]: Child 509 (sleep) died (code=killed, status=9/KILL)
[    8.304979] systemd[1]: Received SIGCHLD from PID 508 (n/a).
[    8.305213] systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 40 (stored), closing.
[    8.305440] systemd[1]: Bus private-bus-connection: changing state RUNNING → CLOSING
[    8.305637] systemd[1]: Bus private-bus-connection: changing state CLOSING → CLOSED
[    8.305832] systemd[1]: Got disconnect on private connection.
[    8.306084] systemd[1]: four.service: Control group is empty.
[    8.306295] systemd[1]: four.service: Deactivated successfully.
[    8.306525] systemd[1]: four.service: Service will not restart (restart setting)
[    8.306735] systemd[1]: four.service: Changed running -> dead
...
[    8.319112] testsuite-56.sh[503]: Finished with result: success
[    8.319112] testsuite-56.sh[503]: Service runtime: 1.000s
[    8.320500] systemd[1]: four.service: Collecting.
[    8.320650] testsuite-56.sh[466]: + echo 'unexpected success'
[    8.321452] testsuite-56.sh[466]: unexpected success
[    8.321452] testsuite-56.sh[466]: + exit 1
```

Reply:
Hmm, that looks like a bug in the core logic? We kill everything with SIGKILL but the service is successful. @poettering PTAL.

Locally I get:
```
[97177.479848] krowka systemd[1]: Started four.service.
[97197.305736] krowka systemd[1]: four.service: Sent signal SIGKILL to main process 1133462 (bash) on client request.
[97197.311401] krowka systemd[1]: four.service: Sending signal SIGKILL to process 1133463 (sleep) on client request.
[97197.312886] krowka systemd[1]: four.service: Last process exited, code=killed, status=9/KILL
[97197.314874] krowka systemd[1]: four.service: Failed with result 'signal'.
```

Reply:
Hmm, I was running this in a loop (with shorter timeouts), and after a couple hundred iterations, I always get `Failed with result 'signal'.`.

Reply:
Hmpf, this is puzzling. I checked the Arch run as well, and there it fails in a different service (eight.service), but with the same issue:

```
[   10.679416] systemd[1]: Started /tmp/test56-exit-cgroup-unclean.sh.
[   10.799063] kernel: audit: type=1130 audit(1617877761.283:108): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=eight comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[   10.679602] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/eight_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=250 reply_cookie=0 signature=sa{s>
[   10.679663] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/eight_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=251 reply_cookie=0 signature=sa{s>
[   10.679729] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=252 reply_cookie=0 signature=uoss error-name=n/a error-messa>
[   10.679784] systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=253 reply_cookie=0 signature=s error-name=n/a error-m>
[   10.679877] systemd[283]: eight.service: Executing: /tmp/test56-exit-cgroup-unclean.sh
[   10.680333] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/eight_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=254 reply_cookie=0 signature=sa{s>
[   10.680395] testsuite-56.sh[282]: Running as unit: eight.service
[   10.680686] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/eight_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=255 reply_cookie=0 signature=sa{s>
[   10.680744] systemd[1]: systemd-journald.service: Got notification message from PID 184 (FDSTORE=1)
[   10.680809] systemd[1]: systemd-journald.service: Added fd 47 (n/a) to fd store.
[   10.680866] systemd[1]: Got message type=method_call sender=:1.9 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/eight_2eservice interface=org.freedesktop.DBus.Properties member=GetAll cookie=5 reply_cookie=0 s>
[   10.680921] systemd[1]: Failed to read pids.max attribute of cgroup root, ignoring: No data available
[   10.681725] systemd[1]: Found unit eight.service at /run/systemd/transient/eight.service (regular file)
[   10.681887] systemd[1]: Preset files don't specify rule for eight.service. Enabling.
[   10.681975] systemd[1]: Sent message type=method_return sender=n/a destination=:1.9 path=n/a interface=n/a member=n/a cookie=256 reply_cookie=5 signature=a{sv} error-name=n/a error-message=n/a
[   10.682555] test56-exit-cgroup-unclean.sh[284]: + sleep 1
[   10.682927] systemd[1]: Received SIGCHLD from PID 283 (bash).
[   10.683013] systemd[1]: Child 283 (bash) died (code=exited, status=0/SUCCESS)
[   10.683073] systemd[1]: eight.service: Child 283 belongs to eight.service.
[   11.684148] test56-exit-cgroup-unclean.sh[284]: + exit 255
[   11.684513] systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 47 (stored), closing.
[   11.684788] systemd[1]: eight.service: Control group is empty.
[   11.684852] audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=eight comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[   11.684934] systemd[1]: eight.service: Deactivated successfully.
[   11.685016] systemd[1]: eight.service: Service will not restart (restart setting)
[   11.685095] systemd[1]: eight.service: Changed running -> dead
[   11.685151] systemd[1]: eight.service: Consumed 5ms CPU time.
[   11.685210] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/eight_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=257 reply_cookie=0 signature=sa{s>
[   11.685273] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/eight_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=258 reply_cookie=0 signature=sa{s>
[   11.685337] systemd[1]: Received SIGCHLD from PID 284 (bash).
[   11.685387] systemd[1]: Child 284 (bash) died (code=exited, status=255/EXCEPTION)
[   11.685437] systemd[1]: system.slice: Child 284 belongs to system.slice.
[   11.685819] systemd[1]: Got message type=method_call sender=:1.9 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/eight_2eservice interface=org.freedesktop.DBus.Properties member=GetAll cookie=6 reply_cookie=0 s>
[   11.685972] systemd[1]: Failed to read pids.max attribute of cgroup root, ignoring: No data available
[   11.804694] kernel: audit: type=1131 audit(1617877762.289:109): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=eight comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[   11.686355] systemd[1]: Sent message type=method_return sender=n/a destination=:1.9 path=n/a interface=n/a member=n/a cookie=259 reply_cookie=6 signature=a{sv} error-name=n/a error-message=n/a
[   11.687147] testsuite-56.sh[282]: Finished with result: success
[   11.687147] testsuite-56.sh[282]: Service runtime: 1.005s
[   11.687147] testsuite-56.sh[282]: CPU time consumed: 5ms
[   11.687647] systemd[1]: Got message type=signal sender=org.freedesktop.DBus destination=n/a path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameOwnerChanged cookie=57 reply_cookie=0 signature=sss error-name=n/a error-m>
[   11.687715] testsuite-56.sh[218]: + echo 'unexpected success'
[   11.687715] testsuite-56.sh[218]: unexpected success
[   11.687715] testsuite-56.sh[218]: + exit 1
```

Reply:
And just managed to reproduce it on CentOS 8 after running the test a couple of times (with this PR):

```
[15394.728851] test56-exit-cgroup.sh[62]: + sleep infinity
[15395.715524] testsuite-56.sh[55]: + systemctl kill --signal 9 four
[15395.729101] systemd[1]: four.service: Sent signal SIGKILL to main process 58 (bash) on client request.
[15395.729220] systemd[1]: four.service: Sending signal SIGKILL to process 59 (sleep) on client request.
[15395.729326] systemd[1]: four.service: Sending signal SIGKILL to process 60 (bash) on client request.
[15395.729435] systemd[1]: four.service: Sending signal SIGKILL to process 61 (sleep) on client request.
[15395.729533] systemd[1]: four.service: Sending signal SIGKILL to process 62 (sleep) on client request.
[15395.730096] systemd[1]: Received SIGCHLD from PID 58 (bash).
[15395.730234] systemd[1]: Child 58 (bash) died (code=killed, status=9/KILL)
[15395.730340] systemd[1]: four.service: Child 58 belongs to four.service.
[15395.730457] systemd[1]: Child 60 (bash) died (code=killed, status=9/KILL)
[15395.730554] systemd[1]: Received SIGCHLD from PID 60 (n/a).
[15395.736205] systemd[1]: Received SIGCHLD from PID 61 (sleep).
[15395.736384] systemd[1]: Child 61 (sleep) died (code=killed, status=9/KILL)
[15395.736532] systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 37 (stored), closing.
[15395.736681] systemd[1]: four.service: Control group is empty.
[15395.736828] systemd[1]: four.service: Deactivated successfully.
[15395.736973] systemd[1]: four.service: Service will not restart (restart setting)
[15395.737144] systemd[1]: four.service: Changed running -> dead
[15395.738241] testsuite-56.sh[56]: Finished with result: success
[15395.738241] testsuite-56.sh[56]: Service runtime: 1.014s
[15395.738990] testsuite-56.sh[27]: + echo 'unexpected success'
[15395.738990] testsuite-56.sh[27]: unexpected success
[15395.738990] testsuite-56.sh[27]: + exit 1
```

Reply:
Oh well, and now neither I nor the CI can reproduce it anymore :shrug:

Reply:
Sounds like a race. the systemd-run might need more than 1s to get things started on an overloaded CI machine and the SIGKILL then already fired?

Reply:
It's sending SIGKILL to 5 different processes, so the service is certainly already up by the time we start killing. And when the tests passes (see my comment above), we also kill 5 processes. So this means that the script has executed successfully and all children have been started.